src/RadixDlt.NetworkGateway.PostgresIntegration/Services/KeyValueStoreQuerier.cs

```diff
@@ -98,68 +97,70 @@ public KeyValueStoreQuerier(IDapperWrapper dapperWrapper, ReadOnlyDbContext dbCo
         _networkConfigurationProvider = networkConfigurationProvider;
     }
 
-    public async Task<GatewayApiSdk.Model.StateKeyValueStoreKeysResponse> KeyValueStoreItems(
+    public async Task<GatewayApiSdk.Model.StateKeyValueStoreKeysResponse> KeyValueStoreKeys(
         EntityAddress keyValueStoreAddress,
         GatewayApiSdk.Model.LedgerState ledgerState,
-        int offset,
-        int limit,
+        GatewayModel.StateKeyValueStoreKeysCursor? cursor,
+        int pageSize,
         CancellationToken token = default)
     {
         var keyValueStore = await GetEntity<InternalKeyValueStoreEntity>(keyValueStoreAddress, ledgerState, token);
         var keyValueStoreSchema = await GetKeyValueStoreSchema(keyValueStore.Id, ledgerState, token);
 
         var cd = new CommandDefinition(
             commandText: @"
-WITH key_value_store_items_slices AS (
-    SELECT key_value_store_entry_ids[@startIndex:@endIndex] AS key_value_store_items_slice, cardinality(key_value_store_entry_ids) AS key_value_store_items_total_count
-    FROM key_value_store_aggregate_history
-    WHERE key_value_store_entity_id = @keyValueStoreEntityId AND from_state_version <= @stateVersion
+SELECT d.id AS Id, d.key AS Key, d.from_state_version AS FromStateVersion, h.value AS Value, h.is_deleted AS IsDeleted, h.is_locked AS IsLocked, h.from_state_version AS LastUpdatedAtStateVersion
+FROM key_value_store_entry_definition d
+INNER JOIN LATERAL (
+    SELECT *
+    FROM key_value_store_entry_history
+    WHERE key_value_store_entry_definition_id = d.id AND from_state_version <= @stateVersion
     ORDER BY from_state_version DESC
     LIMIT 1
-)
-SELECT
-    kvseh.from_state_version AS FromStateVersion,
-    kvseh.key AS Key,
-    kvsits.key_value_store_items_total_count AS TotalCount
-FROM key_value_store_items_slices AS kvsits
-INNER JOIN LATERAL UNNEST(key_value_store_items_slice) WITH ORDINALITY AS key_value_store_join(id, ordinality) ON TRUE
-INNER JOIN key_value_store_entry_history kvseh ON kvseh.id = key_value_store_join.id AND kvseh.is_deleted = FALSE
-ORDER BY key_value_store_join.ordinality ASC
+) h ON TRUE
+WHERE
+    d.key_value_store_entity_id = @keyValueStoreEntityId
+  AND (d.from_state_version, d.id) <= (@cursorStateVersion, @cursorId)
```

Comment:
I'm not sure but maybe providing cursorStateVersion and cursorId as nullable to sql and rework to not apply filters if they are null will be more performant?

Reply:
It results in extra `index cond` (`Filter: (ROW(from_state_version, id) <= ROW('11111111111'::bigint, '2222222222222'::bigint))"`) that should have negligible impact on the query.